x-pack/platform/plugins/shared/stack_alerts/common/esql_query_utils.ts

```diff
@@ -37,27 +51,235 @@ export const rowToDocument = (columns: EsqlResultColumn[], row: EsqlResultRow):
+  const hits: EsqlHit[] = [];
+  const rows: EsqlDocument[] = [];
+  for (const row of table.values) {
+    const document = rowToDocument(table.columns, row);
+    hits.push({
       _id: ESQL_DOCUMENT_ID,
       _index: '',
       _source: document,
-    };
-  });
+    });
+    rows.push(rows.length > 0 ? document : { [ALERT_ID_COLUMN]: 'query matched', ...document });
+  }
 
   return {
-    hits,
-    total: hits.length,
+    results: {
+      isCountAgg: true,
+      isGroupAgg: false,
+      esResult: {
+        took: 0,
+        timed_out: false,
+        _shards: { failed: 0, successful: 0, total: 0 },
+        hits: {
+          hits,
+          total: hits.length,
+        },
+      },
+    },
+    duplicateAlertIds: new Set<string>(),
+    rows,
+    cols: getColumnsForPreview(table.columns),
   };
 };
 
-export const transformDatatableToEsqlTable = (results: Datatable): EsqlTable => {
-  const columns: EsqlResultColumn[] = results.columns.map((c) => ({
+export const toGroupedEsqlQueryHits = (table: EsqlTable, alertId: string[]): EsqlQueryHits => {
+  const duplicateAlertIds: Set<string> = new Set<string>();
+  const rows: EsqlDocument[] = [];
+  const groupedHits = table.values.reduce<Record<string, EsqlHit[]>>((acc, row) => {
+    const document = rowToDocument(table.columns, row);
+    const id = alertId
+      .filter((a) => !isNil(document[a]))
+      .map((a) => document[a])
+      .join(',');
+
+    if (id) {
+      const hit = {
+        _id: ESQL_DOCUMENT_ID,
+        _index: '',
+        _source: document,
+      };
+      if (acc[id]) {
+        duplicateAlertIds.add(id);
+        acc[id].push(hit);
+      } else {
+        acc[id] = [hit];
+      }
+      rows.push({ [ALERT_ID_COLUMN]: id, ...document });
+    }
+    return acc;
+  }, {});
+
+  const aggregations = {
+    groupAgg: {
+      buckets: entries(groupedHits).map(([key, value]) => {
+        return {
+          key,
+          doc_count: value.length,
+          topHitsAgg: {
+            hits: {
+              hits: value,
+            },
+          },
+        };
+      }),
+    },
+  };
+
+  return {
+    results: {
+      isCountAgg: false,
+      isGroupAgg: true,
+      esResult: {
+        took: 0,
+        timed_out: false,
+        _shards: { failed: 0, successful: 0, total: 0 },
+        hits: { hits: [] },
+        aggregations,
+      },
+    },
+    duplicateAlertIds,
+    rows,
+    cols: getColumnsForPreview(table.columns),
+  };
+};
+
+export const transformDatatableToEsqlTable = (datatable: Datatable): EsqlTable => {
+  const columns: EsqlResultColumn[] = datatable.columns.map((c) => ({
     name: c.id,
     type: c.meta.type,
   }));
-  const values: EsqlResultRow[] = results.rows.map((r) => Object.values(r));
+  const values: EsqlResultRow[] = datatable.rows.map((r) => Object.values(r));
   return { columns, values };
 };
+
+export const getAlertId = (query: string, resultColumns: EsqlResultColumn[]): string[] => {
```

Comment:
I'd like to get some feedback on this part of the code, maybe there is a better way to determine the `alertId`?